### PR TITLE
MET-181: add notion of bidirectional association type

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**
The goal of this PR is to add possibility to define an two way association type on creation association type interface. 

The user can show this configuration on edit association type interface and datagrind view.

We also add this information on consult/create/list/upsert API and on import and export file.

Functional rule associated : 
An two way association type cannot be changed to one way association type and vice versa

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Done
| Added legacy Behats               | -
| Added acceptance tests            | Done
| Added integration tests           | Done
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml [hunk 1]
```diff
@@ -11,6 +11,9 @@ datagrid:
             label:
                 label: Label
                 frontend_type: label
+            isBidirectional:
```
Comment (tseho) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml:14]: should we keep `isBidirectional` or simply use `bidirectional` ?

Comment (tseho) [reply src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml:14]: That would simplify the thing with camelcase/snakecase

Comment (tseho) [reply src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml:14]: or `two_way_association` ?

File: upgrades/schema/Version_5_0_20200320095430_add_is_bidirectional_col_to_associations.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version_5_0_20200320095430_add_is_bidirectional_col_to_associations extends AbstractMigration
+{
+    public function up(Schema $schema) : void
+    {
+        $this->addSql('ALTER TABLE pim_catalog_association_type ADD isBidirectional TINYINT(1) DEFAULT \'0\' NOT NULL');
```
Comment (tseho) [inline upgrades/schema/Version_5_0_20200320095430_add_is_bidirectional_col_to_associations.php:15]: I would rather have a snakecase column name ...

File: tests/legacy/features/pim/structure/association-type/import_association_types.feature [hunk 3]
```diff
@@ -7,14 +7,14 @@ Feature: Import association types
     Given the "footwear" catalog configuration
     And the following CSV file to import:
       """
-      code;is_bidirectional;label-en_US;label-fr_FR
+      code;is_two_way;label-en_US;label-fr_FR
```
Comment (tseho) [inline tests/legacy/features/pim/structure/association-type/import_association_types.feature:10]: `is_two_way` or `two_way` there ?

Comment (StevenVAIDIE) [reply tests/legacy/features/pim/structure/association-type/import_association_types.feature]: Fixed in : 5982a705821ebea7caec3b7c7c058899726a0602

File: src/Akeneo/Pim/Structure/Component/Normalizer/Standard/AssociationTypeNormalizer.php [hunk 4]
```diff
@@ -26,12 +26,14 @@ public function __construct(NormalizerInterface $translationNormalizer)
...
     /**
      * {@inheritdoc}
+     * @param AssociationTypeInterface $associationType
      */
     public function normalize($associationType, $format = null, array $context = [])
     {
         return [
             'code'   => $associationType->getCode(),
             'labels' => $this->translationNormalizer->normalize($associationType, 'standard', $context),
+            'two_way' => $associationType->isTwoWay()
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Component/Normalizer/Standard/AssociationTypeNormalizer.php:36]:
```suggestion
            'is_two_way' => $associationType->isTwoWay()
```

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml [hunk 5]
```diff
@@ -11,6 +11,9 @@ datagrid:
             label:
                 label: Label
                 frontend_type: label
+            two_way:
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/association_type.yml:14]:
```suggestion
            is_two_way:
```

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/create.yml [hunk 6]
```diff
@@ -15,6 +15,17 @@ extensions:
         module: pim/form/common/creation/field
         parent: pim-association-type-create-modal
         targetZone: fields
+        position: 10
         config:
             identifier: code
             label: pim_common.code
+
+    pim-association-type-properties-two-way:
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/create.yml:23]:
```suggestion
    pim-association-type-properties-is-two-way:
```

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/create.yml [hunk 7]
```diff
@@ -15,6 +15,17 @@ extensions:
...
+        module: pim/form/common/fields/boolean
+        parent: pim-association-type-create-modal
+        targetZone: fields
+        position: 20
+        config:
+            fieldName: two_way
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/create.yml:29]:
```suggestion
            fieldName: is_two_way
```

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/edit.yml [hunk 8]
```diff
@@ -107,6 +107,16 @@ extensions:
           formRequired: pim_common.required_label
           inputField: pim_enrich_association_type_form_code
 
+    pim-association-type-edit-form-properties-is-two-way:
+        module: pim/form/common/fields/boolean
+        parent: pim-association-type-edit-form-properties
+        targetZone: accordion
+        position: 110
+        config:
+            fieldName: two_way
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/association_type/edit.yml:116]: same

File: src/Akeneo/Tool/Component/Api/tests/integration/Normalizer/AssociationTypeIntegration.php [hunk 9]
```diff
@@ -17,6 +17,7 @@ public function testAssociationType()
                 'en_US' => 'Cross sell',
                 'fr_FR' => 'Vente croisée',
             ],
+            'two_way' => false
```
Comment (juliensnz) [inline src/Akeneo/Tool/Component/Api/tests/integration/Normalizer/AssociationTypeIntegration.php:20]: same

File: src/Akeneo/Pim/Structure/Component/Normalizer/Standard/AssociationTypeNormalizer.php (hunk 4 above)
Comment (StevenVAIDIE) [reply src/Akeneo/Pim/Structure/Component/Normalizer/Standard/AssociationTypeNormalizer.php:36]: Actually, the field scopable/localizable/unique/required on attribute are stocked in database in format is_scopable/is_localizable/is_unique/is_required and are exposed in external API as scopable/localizable/unique/required I following this when I wrote code. Should I replace all by is_two_way ?

**Comment by tseho** [comment]
Regarding the naming, I've been looking at the existing code in the PIM.
Let's take the booleans on `Attribute`:

Screenshot_2020-04-09_11-44-24

https://github.com/akeneo/pim-community-dev/blob/cf2bb77a0d86a695e2d8914e4bc8833b475159d8/src/Akeneo/Pim/Structure/Bundle/Resources/config/model/doctrine/Attribute.orm.yml

Screenshot_2020-04-09_11-45-22

Screenshot_2020-04-09_11-46-19

https://github.com/akeneo/pim-community-dev/blob/cf2bb77a0d86a695e2d8914e4bc8833b475159d8/src/Akeneo/Pim/Structure/Component/Model/AbstractAttribute.php

And in the *API*:

Screenshot_2020-04-09_11-47-21

https://api.akeneo.com/api-reference.html#get_attributes

---

What is currently done is:
column name: `is_x`
entity property: `x`
entity getter: `isX()`
api property: `x`

I suggested to steven to follow how it's already done but it's clearly up to discussion, we don't want something confusing.

**Pull request merged**